### DISCUSSION TITLE
Refactoring utility.py and search.py

### DISCUSSION THREAD

**Discussion by NamamiShanker** [IndianOpenSourceFoundation/dynamic-cli#101] [Ideas]
I think utility.py file has become very big and a bit confusing. It will be a bit difficult for new contributors to understand its functionality easily. It would be better if we split it into utility.py, questionpanel.py, playbook.py or some other structure. We can shift classes and improve existing implementation.

I think the current structure of search.py is also messy.

We should discuss this and share some ideas.

**Comment by GouravSardana** [comment]
Yes code refactoring is necessary. I agree with your point. 
I suggest @pranavbaburaj to move all the test file to a new folder called test as well.

I think we can make a new folder called as features and make file called as api.py (all the request like get,put,post), playbook.py, notion.py(logging into notion to saving the data into notion), display.py, search.py

What do you guys suggest ?
@guptaaastha @pranavbaburaj @Siddhant-K-code @anshsrtv ?

**Comment by lainq** [comment]
Yes, I think the project structure is a bit messed up. All the source files and included in the `src/arguments`directory.This directory contains code that's not even related to arguments. I think we have to split them into different modules by adding an `__init__.py` file.

```ps1
[src]
- error : # error files
- tests: # for test files
- ext : # other files ,like the update.py files
- core : # all the core stuff
```

Please let me know your thoughts

**Reply by NamamiShanker** [reply]
I think this will work! Waiting for others to reply.

**Comment by Siddhant-K-code** [comment]
Yes, Let's Start with the test folder and then we can try to optimize and write clean code for it. I think first, we should plan what changes we should do, before implementing it.

**Reply by lainq** [reply]
Yeah, we should write some tests. We need to have solid tests. We can also set up a workflow that runs all the tests. What do you guys think 🤔 ??

**Reply by NamamiShanker** [reply]
Yeah! There is an inactive issue regarding tests. I think the main contributors should design the test module. They are the one's most familiar with the project.

**Reply by GouravSardana** [reply]
we can go with @pranavbaburaj structure which he suggested above

**Comment by NamamiShanker** [comment]
So how do we go about doing this? Should we make a new branch and refactor the code with it?

**Reply by GouravSardana** [reply]
Yes, we can make a new branch i.e dev. You guys can make your own branch to work on it but make sure while you sending the PR, target should be iosf dev branch.

**Reply by guptaaastha** [reply]
I agree, refactor in a different branch. My suggestion would be to pick up anyone first (either uitility.py or search.py) and start by sections instead of doing it all together to avoid confusion and add relevant comments wherever necessary 😄

**Reply by GouravSardana** [reply]
Pick search first then will move on to utilities then errors

**Reply by GouravSardana** [reply]
Just to give you an update- Pranav account got suspended for no reason. He also didn't know the reason. He created a ticket for GitHub support. It takes time around 1-2 days. That's why the thread that was created by Pranav got removed.

cc @anshsrtv @guptaaastha @NamamiShanker @Siddhant-K-code

**Reply by lainq** [reply]
Hey, everyone. I got my account back. Thank you @GouravSardana for your help and support :)

**Reply by GouravSardana** [reply]
@pranavbaburaj is giving virtual party for his account 🎈🎉
